### PR TITLE
debconf: add 'reconfigure' option

### LINKED ISSUE #29352: Debconf to (be able to) perform dpkg-reconfigure
From @tkteun on 2016-08-22T12:15:29Z
##### ISSUE TYPE

Feature Idea
##### COMPONENT NAME

debconf
##### ANSIBLE VERSION

```
ansible 2.1.1.0
```
##### CONFIGURATION

N/A
##### OS / ENVIRONMENT

Debian 8.5 Jessie
##### SUMMARY

debconf only sets configuration parameters, but can't execute/set them.
##### STEPS TO REPRODUCE

```
- name: Enable unattended security updates
  debconf: name=unattended-upgrades
    question="unattended-upgrades/enable_auto_updates"
    value=true
    vtype=boolean
    reconfigure=true
```
##### EXPECTED RESULTS

```
- name: Enable unattended security updates
  debconf: name=unattended-upgrades
    question="unattended-upgrades/enable_auto_updates"
    value=true
    vtype=boolean

- name: Reconfigure unattended upgrades with dpkg
  command: '/usr/sbin/dpkg-reconfigure --frontend noninteractive unattended-upgrades'
```
##### ACTUAL RESULTS

Variable is being set in debconf, but it's not actually applied to the package.

Copied from original issue: ansible/ansible-modules-extras#2765

### PR DESCRIPTION
**Pull request by smeso**
##### SUMMARY
Add a new 'reconfigure' option to 'debconf' module.
Ansible will run 'dpkg-reconfigure' against the given package whenever:
  - a change is detected
  - the package is installed
  - 'reconfigure' is set to true

In the process of implementing this feature a new module utility for APT
is created. This way we can avoid code duplication across different
APT-related modules.

Fixes #29352

##### ISSUE TYPE
 - Feature Pull Request
 - New Module Pull Request

##### COMPONENT NAME
debconf
apt
apt_repository
apt_utils

### CODE REVIEW

**Comment by smeso** [comment]
Shippable failed a test on freebsd with `ERROR: Maximum threshold reached and all endpoints exhausted.` which seems completely unrelated to my code. 
Especially since this PR is about APT and debconf and so it's Debian-only.

ready_for_review

**Comment by smeso** [comment]
ready_for_review

File: lib/ansible/module_utils/apt_utils.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
+# -*- coding: utf-8 -*-
+
+# Copyright: (c) 2012, Flowroute LLC
+# Copyright: (c) 2018, Salvatore Mesoraca <s.mesoraca16@gmail.com>
+# Based on apt module written by Matthew Williams <matthew@flowroute.com>
+
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```
Comment (bcoca) [inline lib/ansible/module_utils/apt_utils.py:7]: we really prefer to use BSD on module_utils for 3rd party module distribution, if any of my code is involved, this also counts as my agreement to relicense

**Review by bcoca — CHANGES_REQUESTED** [review]

File: lib/ansible/module_utils/apt_utils.py (hunk 1 above)
Comment (smeso) [reply lib/ansible/module_utils/apt_utils.py:7]: I have no problem doing this, if @mgwilliams agrees.

Comment (smeso) [reply lib/ansible/module_utils/apt_utils.py:7]: While I was waiting for a response from Matthew Williams, I realized that it isn't possible to re-license this module. It uses python-apt, which is GPL-2+ (http://changelogs.ubuntu.com/changelogs/pool/main/p/python-apt/python-apt_1.6.0/copyright) and so it can't be released with a BSD license.

**Comment by smeso** [comment]
I realized that it isn't possible to re-license this module. It uses python-apt, which is GPL-2+ (http://changelogs.ubuntu.com/changelogs/pool/main/p/python-apt/python-apt_1.6.0/copyright) and so it can't be released with a BSD license.

ready_for_review

**Comment by abadger** [comment]
It is okay to license this code as BSD.  When the code is used in a program, it falls under the terms of the GPL because python-apt is licensed under the GPL.  However, when someone copies this code (for instance, as a framework for a module_util for a different package management system), the copied code can be used under the BSD license the libraries for that package management system allow that.

**Comment by smeso** [comment]
@abadger  I'm not a legalese expert. Do you know what would be the best way to state that the code is released under BSD license, without hiding the fact that, when used, it falls under the terms of the GPL?
Is there any snippet that I could just copy/paste?

**Comment by abadger** [comment]
It looks like no one from the meeting pushed this to someone who talks to lawyers.  I've given this to our managers to pose to the Red Hat lawyers now.

**Comment by smeso** [comment]
Thank you very much, I'll wait for a response.

**Comment by bcoca** [comment]
@smeso the GPL in module_utils topic was brought up in last IRC meeting, we voted to make an exception for narrowly focused module sets. @abadger do we agree this qualifies?

**Review by bcoca — APPROVED** [review]

**Comment by tgadiev** [comment]
Is there any progress?

**Comment by etursunbaev** [comment]
Yes it would be nice to have this option.

**Comment by rako1980** [comment]
Looks like issue has been addressed, but hung up on conflict?  @bcoca

**Comment by seanporterwork** [comment]
Any movement on this in the last six months?
Adding my +1 to this being useful...

**Comment by umataro** [comment]
+1

**Comment by RCM7** [comment]
+1

**Comment by SkullTech** [comment]
+1 this would be a real nice-to-have.

File: lib/ansible/module_utils/apt_utils.py (hunk 1 above)
Comment (abadger) [reply lib/ansible/module_utils/apt_utils.py:7]: Legal just updated me so I'm going to update this comment thread (even though I don't know if this PR will be worked on in the future or not):

It is fine for this file to be licensed under the BSD if the authors agree to that.  The file itself is BSD in case someone ever extracts the code to use in a different context or in case ideas from the code were to be copied to another module_utils file.  When used together with the python-apt library, the work as a whole is used under the terms of the GPLv2+.

Legal does recommend that when cases like this arise that have the potential to confuse the users of the code, an explicit comment is desirable to warn of the confusion.  For instance, adding a comment like the following right after the BSD license header:

``` python
# Note: Although this code is BSD-licensed, it imports a module that is licensed under the GPL.
```

**Comment by lbt** [comment]
TL;DR; :  running dpkg-reconfigure may work sometimes but it's very unreliable.


I was struggling with this and trying to find a workaround and thinking of some docs for the module. I began writing these which I hope to create a PR with when I'm done.

--- Potential documentation ---

Notes: Reconfiguring packages in Debian using debconf is not straightforward.

The debconf module does not reconfigure packages, it just updates the debconf database. An additional step is needed (typically via notify if debconf makes a change) to reconfigure the package and apply the changes.

debconf is primarily used for pre-seeding configuration prior to installation.

So, whilst dpkg-reconfigure does use debconf data, it is not always authoritative and you may need to check how your package is handled.

dpkg-reconfigure is a 3-phase process. It invokes the control scripts from the /var/lib/dpkg/info directory with the following arguments:
  <package>.prerm  reconfigure <version>
  <package>.config reconfigure <version>
  <package>.postinst control <version>

The main issue is that the "<package>.config reconfigure" step for many packages will first reset the debconf database (overriding changes made by this module) by checking the on-disk configuration. If this is the case for your package then dpkg-reconfigure will effectively ignore changes made by this debconf module.

However as dpkg-reconfigure finally invokes:
    /var/lib/dpkg/info/<package>.postinst configure <version>
to actually configure the package this may be a suitable trigger.
<version> is dpkg --status <package> | grep "^Version:" | cut -f2 -d" "
Use with care.

--- End of potential documentation ---


That final part turns out not to be that simple. The script is expected to be run from a frontend and uses IPC to respond to the _db_cmd statements in the script.

To see this in more detail
export DPKG_MAINTSCRIPT_PACKAGE=<package>
export DPKG_MAINTSCRIPT_NAME=<script path>
export DEBIAN_HAS_FRONTEND=1

and run the script. I was trying to setup unattended-upgrades so I ran:

sh -x /var/lib/dpkg/info/unattended-upgrades.postinst configure 1.11.2

**Comment by lbt** [comment]
I can confirm that running `/var/lib/dpkg/info/unattended-upgrades.postinst configure 1.11.2` sources `/usr/share/debconf/confmodule` which exec()s `/usr/share/debconf/frontend` which forces the `<package>.config configure` phase to take place.

This is based on the existence of the .config file and can only be avoided by mv'ing `/var/lib/dpkg/info/<package>.config` out of the way whilst dpkg-reconfigure (or other related debconf code) runs. This is probably worth doing in the python reconfigure option code.

The programmers manual says that the config script's sole purpose is to populate debconf and must not affect other files so this should be compliant with debian policy: http://www.fifi.org/doc/debconf-doc/tutorial.html#AEN113

**Comment by Akasurde** [comment]
@gundalow @abadger What can we do to get this PR rolling? Let me know, happy to help.

**Comment by s-hertel** [comment]
This PR can't be merged until the license is fixed.
